relations/src/preimage.rs

```diff
@@ -0,0 +1,141 @@
+// This relation showcases how to use Poseidon in r1cs circuits
+use ark_bls12_381::Fr;
+use ark_r1cs_std::{alloc::AllocVar, eq::EqGadget};
+use ark_relations::{
+    ns,
+    r1cs::{
+        ConstraintSynthesizer, ConstraintSystemRef, SynthesisError,
+        SynthesisError::AssignmentMissing,
+    },
+};
+use ark_std::{marker::PhantomData, vec, vec::Vec};
+use once_cell::sync::Lazy;
+use poseidon::r1cs;
+
+use crate::{
+    environment::FpVar,
+    relation::state::{FullInput, NoInput, OnlyPublicInput, State, WithPublicInput},
+    CircuitField, GetPublicInput,
+};
+
+// Poseidon paper suggests using domain separation for this, concretely encoding the use case in the capacity element (which is fine as it is 256 bits large and has a lot of bits to fill)
+static DOMAIN_SEPARATOR: Lazy<Fr> = Lazy::new(|| Fr::from(2137));
+
+/// Preimage relation : H(preimage)=hash
+/// where:
+/// - hash : public input
+/// - preimage : private witness
+#[derive(Copy, Clone, Eq, PartialEq, Hash, Debug)]
+pub struct PreimageRelation<S: State> {
+    // private witness
+    pub preimage: Option<CircuitField>,
+    // public input
+    pub hash: Option<CircuitField>,
+    _phantom: PhantomData<S>,
+}
+
+impl PreimageRelation<NoInput> {
+    pub fn without_input() -> Self {
+        Self {
+            hash: None,
+            preimage: None,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl PreimageRelation<OnlyPublicInput> {
+    pub fn with_public_input(hash: CircuitField) -> Self {
+        Self {
+            preimage: None,
+            hash: Some(hash),
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl PreimageRelation<FullInput> {
+    pub fn with_full_input(preimage: CircuitField, hash: CircuitField) -> Self {
+        Self {
+            preimage: Some(preimage),
+            hash: Some(hash),
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<S: State> ConstraintSynthesizer<CircuitField> for PreimageRelation<S> {
+    fn generate_constraints(
+        self,
+        cs: ConstraintSystemRef<CircuitField>,
+    ) -> Result<(), SynthesisError> {
+        let preimage = FpVar::new_witness(ns!(cs, "preimage"), || {
+            self.preimage.ok_or(AssignmentMissing)
+        })?;
+        let hash = FpVar::new_input(ns!(cs, "hash"), || self.hash.ok_or(AssignmentMissing))?;
+
+        let domain_separator = FpVar::new_constant(cs.clone(), *DOMAIN_SEPARATOR)?;
+        let hash_result = r1cs::one_to_one_hash(cs, &domain_separator, preimage)?;
+
+        hash.enforce_equal(&hash_result)?;
+
+        Ok(())
+    }
+}
+
+impl<S: WithPublicInput> GetPublicInput<CircuitField> for PreimageRelation<S> {
+    fn public_input(&self) -> Vec<CircuitField> {
+        vec![self
+            .hash
+            .expect("Circuit should have public input assigned")]
+    }
+}
+
+#[cfg(test)]
```

Comment:
I'd add two more tests for negative cases to make sure that `is_satisfied` and `verify` are not implemented as `{ true }`

Reply:
if it did, that would be very unfortunate :)

I created a test where constraints are unsatisfied - imo there is no need to check whether prover works in that case, since they all check the circuit even before creating the proof (see [here](https://github.com/arkworks-rs/groth16/blob/master/src/prover.rs#L253))